klippa/src/serialize.rs

```diff
@@ -0,0 +1,354 @@
+
+use fnv::FnvHashMap;
+use write_fonts::types::Scalar;
+
+#[derive(Clone, Copy, Debug, PartialEq)]
+#[allow(dead_code)]
+pub(crate) struct SerializeErrorFlags(u16);
+
+#[allow(dead_code)]
+impl SerializeErrorFlags {
+    pub const SERIALIZE_ERROR_NONE: Self = Self(0x0000);
+    pub const SERIALIZE_ERROR_OTHER: Self = Self(0x0001);
+    pub const SERIALIZE_ERROR_OFFSET_OVERFLOW: Self = Self(0x0002);
+    pub const SERIALIZE_ERROR_OUT_OF_ROOM: Self = Self(0x0004);
+    pub const SERIALIZE_ERROR_INT_OVERFLOW: Self = Self(0x0008);
+    pub const SERIALIZE_ERROR_ARRAY_OVERFLOW: Self = Self(0x0010);
+}
+
+impl Default for SerializeErrorFlags {
+    fn default() -> Self {
+        Self::SERIALIZE_ERROR_NONE
+    }
+}
+
+impl std::ops::BitOrAssign for SerializeErrorFlags {
+    /// Adds the set of flags.
+    #[inline]
+    fn bitor_assign(&mut self, other: Self) {
+        self.0 |= other.0;
+    }
+}
+
+impl std::ops::Not for SerializeErrorFlags {
+    type Output = bool;
+    #[inline]
+    fn not(self) -> bool {
+        self == SerializeErrorFlags::SERIALIZE_ERROR_NONE
+    }
+}
+
+#[allow(dead_code)]
+#[derive(Default)]
+// Offset relative to the current object head (default)/tail or
+// Absolute: from the start of the serialize buffer
+enum OffsetWhence {
+    #[default]
+    Head,
+    Tail,
+    Absolute,
+}
+
+type PoolIdx = usize;
+// Reference Harfbuzz implementation:
+// <https://github.com/harfbuzz/harfbuzz/blob/5e32b5ca8fe430132b87c0eee6a1c056d37c35eb/src/hb-serialize.hh#L69>
+#[derive(Default)]
+#[allow(dead_code)]
+struct Object {
+    // head/tail: indices of the output buffer for this object
+    head: usize,
+    tail: usize,
+    // real links are associated with actual offsets
+    real_links: Vec<Link>,
+    // virtual links not associated with actual offsets,
+    // they exist merely to enforce an ordering constraint.
+    virtual_links: Vec<Link>,
+
+    // pool index of the object that will be worked on next
+    next_obj: Option<PoolIdx>,
+}
+
+// LinkWidth:2->offset16, 3->offset24 and 4->offset32
+#[allow(dead_code)]
+#[derive(Default)]
+enum LinkWidth {
+    #[default]
+    Two,
+    Three,
+    Four,
+}
+
+type ObjIdx = u32;
+
+#[allow(dead_code)]
+#[derive(Default)]
+struct Link {
+    width: LinkWidth,
+    is_signed: bool,
+    whence: OffsetWhence,
+    bias: u32,
+    position: u32,
+    objidx: Option<ObjIdx>,
+}
+
+// reference harfbuzz implementation:
+//<https://github.com/harfbuzz/harfbuzz/blob/5e32b5ca8fe430132b87c0eee6a1c056d37c35eb/src/hb-serialize.hh#L57>
+#[derive(Default)]
+#[allow(dead_code)]
+pub(crate) struct Serializer {
+    start: usize,
+    end: usize,
+    head: usize,
+    tail: usize,
+    // TODO: zerocopy, debug_depth
+    errors: SerializeErrorFlags,
+
+    data: Vec<u8>,
+    object_pool: ObjectPool,
+    // index for current Object in the object_pool
+    current: Option<PoolIdx>,
+
+    packed: Vec<Option<ObjIdx>>,
```

Comment:
The only item in this vector that will be None is the first one right? If so maybe we should just do the implementation in a way that doesn't require to have a single None object at the start? Then this can just be a Vec<ObjIdx>

Reply:
Agreed. I'll address this in later PRs

Reply:
One other thing when you address this in the later PR: I believe this should be pool index instead of object index.